### PR TITLE
feat(storage): filter data by watermark in fast compact algorithm

### PR DESCRIPTION
**Pull request by Little-Wallace**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/hummock/iterator/skip_watermark.rs [hunk 1]
```diff
@@ -85,7 +170,38 @@ impl<I: HummockIterator<Direction = Forward>> SkipWatermarkIterator<I> {
...
+                    }
+                    Ordering::Greater => {
+                        // The current key has advanced over the watermark.
+                        // We may advance the watermark before advancing the key.
+                    }
+                }
+            } else {
+                return false;
+            }
+            if !self.advance_watermark(key) {
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply src/storage/src/hummock/iterator/skip_watermark.rs]: fixed

**Comment by Little-Wallace** [comment]
For ascending watermark, if the smallest key of a block is larger than watermark key, all keys in this block whose vnode equals the first key shall not be deleted by watermark. But if there are some keys which belong to another vnode, they may need to skip by watermark. 
For descending watermark, if the smallest key of a block is smaller than watermark key, there may be some other keys are larger than watermark key, which means that they should be deleted by watermark but we keep them in output file.


Since watermark will be always monotonically. The data of every vnode of one table will only keep at most one block which should be deleted. It means that for a table with 256 vnode, it will take up to 16MB(256 * 64KB) of additional space because of fast compact algorithm

**Review by Li0k — APPROVED** [review]
Generally LGTM， PTAL @wenym1

**Review by wenym1 — APPROVED** [review]
LGTM!

**Pull request merged**